rclcpp/src/rclcpp/time_source.cpp

```diff
@@ -241,6 +246,11 @@ void TimeSource::create_clock_sub()
       rclcpp::QosPolicyKind::History,
       rclcpp::QosPolicyKind::Reliability,
     });
+  clock_callback_group_ = node_base_->create_callback_group(
```

Comment:
does this need to be initialized every time `use_sim_time` becomes true? i think this should be initialized somewhere else just once and store it as in option here.

Reply:
It is initialized in `create_clock_sub` method, which [automatically returns if the subscription to clock is already created](https://github.com/wyca-robotics/rclcpp/blob/2c6024c8f49cf8c766c8a180cbcda05589b57f2d/rclcpp/src/rclcpp/time_source.cpp#L236). 
In one hand, if the subscription is reset and initialized again, I agree that the callback group shouldn't be initialized again. But the callback group is not useful if `use_sim_time = false`. So it can be initialized only once, but may be not in the class constructor, since it might remain unused.
A solution would be to test in the `create_clock_sub` whether `clock_callback_group_` has already been initialized:
```suggestion
  if (clock_callback_group_ == nullptr) {
    clock_callback_group_ = node_base_->create_callback_group(
      rclcpp::CallbackGroupType::MutuallyExclusive,
      false
    );
  }
```
What do you think about this?

Reply:
I also added [a test on `clock_executor_thread_.joinable()`](https://github.com/wyca-robotics/rclcpp/blob/a82efabf75f0fa78523985eda775815168ed6f18/rclcpp/src/rclcpp/time_source.cpp#L257) to make sure the thread has ended before starting a new one.